packages/auth-construct/API.md

```diff
@@ -20,9 +23,11 @@ import { UserPoolIdentityProviderSamlMetadata } from 'aws-cdk-lib/aws-cognito';
 export type AmazonProviderProps = Omit<aws_cognito.UserPoolIdentityProviderAmazonProps, 'userPool'>;
 
 // @public
-export class AmplifyAuth extends Construct implements ResourceProvider<AuthResources> {
+export class AmplifyAuth extends Construct implements ResourceProvider<AuthResources>, ResourceAccessAcceptorFactory<AuthRoleName> {
     constructor(scope: Construct, id: string, props?: AuthProps);
     addTrigger: (event: TriggerEvent, handler: IFunction | AmplifyFunction) => void;
+    // (undocumented)
+    getResourceAccessAcceptor: (roleName: AuthRoleName) => ResourceAccessAcceptor;

```

Comment:
How are customers using AmplifyAuth CDK construct expected to use this?

Since we are creating CDK constructs, we should stay cdk idiomatic, meaning this should be implmenting a `IGrantable` and Storage should have `grant*` methods. This way these constructs can nicely play with other cdk constructs as well.

Reply:
There are 3 reasons we can't use IGrantable:
1. The Auth construct has multiple roles (currently just the auth and unauth roles, but once this PR lands, there will also be group roles). The IGrantable interface can only wrap a single "grantable entity". So at best we'd have to implement "IGrantableFactory" or something like that
2. We need to pass additional context beyond just a CDK policy. This applies to functions rather than auth, but we still need a consistent interface to handle both of these cases
3. The `grant*` pattern is not flexible enough (or at least not in an intuitive way) for the access control DX that the product team wants. Also, our storage construct has moved to an implementation detail of `backend-storage` so it's not an externally consumable component

While the `ResourceAccessAcceptorFactory<AuthRoleName>` is not CDK idiomatic, it is useable in native CDK. Perhaps I could rename it to `PolicyAcceptorFactory` with a method of `getPolicyFactory`? That at least makes it obvious that you should pass in a policy rather than whatever "ResourceAccess" is. I was trying to name it something that described both the policy and the additional context values, but perhaps the name is more confusing than helpful.

Reply:
1. Can't we do `grant(auth.unauthRole...)`
2. I thought CDK does that. `bucket.grantRead(function);` The `grant*` method as I see it is not prescribed by CDK but up to the implementer. The only requirement is that it should accept a `IGrantable` which our function should be and Roles already are.
3. Let's keep our construct cdk idiomatic and we can do whatever we want in the `define...` factories to get closer to product requirements. Because of product reqs, we can't sacrifice on getting the right CDK constructs out (one of the biggest value proposition of Gen2)

Reply:
Let me play around and see how much of this we can move to the `define*` layer